### PR TITLE
feat(api): add `Image` class

### PR DESCRIPTION
**Pull request by amrbashir**

### CODE REVIEW

**Review by lucasfernog — DISMISSED** [review]

**Review by lucasfernog — DISMISSED** [review]

**Comment by lucasfernog** [comment]
@amrbashir let me know if you agree with my changes :P

File: core/tauri/src/image/plugin.rs [hunk 1]
```diff
@@ -5,46 +5,41 @@
...
-  let rid = resources_table.add(image);
-  Ok(rid)
-}
-#[command(root = "crate")]
-fn from_ico_bytes<R: Runtime>(app: AppHandle<R>, image: JsImage<'_>) -> crate::Result<ResourceId> {
-  let image: Image<'_> = image.try_into()?;
-  let image = image.to_owned();
-  let mut resources_table = app.resources_table();
-  let rid = resources_table.add(image);
+  let rid = resources_table.add(Arc::try_unwrap(image).unwrap().to_owned());
```
Comment (amrbashir) [inline core/tauri/src/image/plugin.rs]: I guess we can just use `Image::new(&rgba, width, height).to_owned()` instead or wrapping in an Arc and unwraping

File: core/tauri/src/image/plugin.rs [hunk 2]
```diff
@@ -5,46 +5,41 @@
...
   Ok(rid)
 }
+
 #[command(root = "crate")]
-fn from_bytes<R: Runtime>(app: AppHandle<R>, image: JsImage<'_>) -> crate::Result<ResourceId> {
-  let image: Image<'_> = image.try_into()?;
-  let image = image.to_owned();
+fn from_bytes<R: Runtime>(app: AppHandle<R>, bytes: Vec<u8>) -> crate::Result<ResourceId> {
+  let image = JsImage::Bytes(&bytes).into_img(&app)?;
```
Comment (amrbashir) [inline core/tauri/src/image/plugin.rs]: ditto

File: core/tauri/src/image/plugin.rs [hunk 3]
```diff
@@ -5,46 +5,41 @@
...
+  let rid = resources_table.add(Arc::try_unwrap(image).unwrap().to_owned());
   Ok(rid)
 }
+
 #[command(root = "crate")]
-fn from_path<R: Runtime>(app: AppHandle<R>, image: JsImage<'_>) -> crate::Result<ResourceId> {
-  let image: Image<'_> = image.try_into()?;
-  let image = image.to_owned();
+fn from_path<R: Runtime>(app: AppHandle<R>, path: std::path::PathBuf) -> crate::Result<ResourceId> {
+  let image = JsImage::Path(path).into_img(&app)?.to_owned();
```
Comment (amrbashir) [inline core/tauri/src/image/plugin.rs]: ditto

**Review by amrbashir — COMMENTED** [review]
Any reason to remove `fromIcoBytes` and `fromPngBytes`?

**Comment by lucasfernog** [comment]
they're the same as fromBytes, we could create aliases on the JS side in case they differ in the future, but for now i don't see a reason to have three separate commands specially since if you allow the from_bytes one you can "bypass" the allowlist pretty much

**Comment by amrbashir** [comment]
> they're the same as fromBytes, we could create aliases on the JS side in case they differ in the future

Well that was just a mistake in my implementation, they should be different as `fromBytes` will use infer to determine the bytes and fromIcoBytes and fromPngBytes should not.

**Review by lucasfernog — APPROVED** [review]

**Pull request merged**